ENH: Integrate to/from wkt/wkb functions into public API

johnflavin opened the pull request:
* Add `GeoSeries.to_wkt`, `GeoSeries.to_wkb` to encode geometries and return pandas `Series`
* Add `GeoSeries.from_wkt`, `GeoSeries.from_wkb` ~~static~~ class methods to construct `GeoSeries` from encoded geometries
* Add `GeoDataFrame.to_wkt`, `GeoDataFrame.to_wkb` to encode all columns in the gdf with type `geometry` and return pandas `DataFrame`
* ~~Add keyword args `wkt` and `wkb` to `GeoDataFrame` constructor. These function similarly to the `geometry` kwarg but first decode the values.~~ Removed. Added example for constructing `GeoDataFrame` with wkb/t by first constructing `GeoSeries`.

Closes #1680
Closes #1689

martinfleis commented:
Thanks a lot for this. I guess that my `good first issue` tag on that issue wasn't the best fit :).

Re your questions from #1680:

> Should `GeoSeries.to_wkt()` / `.to_wkb()` return a pandas Series?

Yes.

> How should `from_wkt` and `from_wkb` be integrated into the API? I could see them being useful as class methods on GeoSeries that would create a GeoSeries from a list or array of WKTs/WKBs. Or I could see them being useful as additional kwargs on the GeoDataFrame constructor; they could work similarly to the geometry kwarg, where a user could specify a wkt or wkb column name or pass in a list of WKTs/WKBs which would be set as the geometry. Or maybe both?

This may require more thought. We have currently only one function of a similar nature - `geopandas.points_from_xy` which stands a bit out of general API. It feels ideal to have these functions exposed as `GeoSeries.from_wkt()` rather than top-level API as the one for points. 

I am less certain about `wkt` and `wkb` kwargs in the `GeoDataFrame` constructor. It can be handy, but I'd like to avoid unnecessary complexity. Also, if you pass `wkt`, you lose your original wkt strings (they are overwritten by geometry). That is an alteration of data users should do themselves. I would keep GeoSeries methods and document their usage in cases when your DataFrame comes with wkt and skip GeoDataFrame stuff. But others may have a different opinion. I am not particularly strongly against it.

johnflavin commented:
> `wkt` and `wkb` kwargs in the GeoDataFrame constructor... can be handy, but I'd like to avoid unnecessary complexity.

I agree with this framing. The `wkt` and `wkb` keyword args are a tradeoff between convenience for users and complexity in the `geopandas` code. I lean more toward favoring convenience for users, but I understand that's an opinion and others might feel differently.

> Also, if you pass `wkt`, you lose your original wkt strings (they are overwritten by geometry). That is an alteration of data users should do themselves.

This is a good point that I hadn't considered. And any ways I can think of right now to avoid overwriting the wkb strings introduce a lot more complexity than the convenience would be worth.

Of course, users could get around this behavior if it isn't what they want. But that's bad because 1. they have to know it is happening first, and if it is unexpected they won't know by definition, and 2. I'm making users jump through hoops to _not_ have their data get manipulated in ways they don't want.

Yeah, this has convinced me to not have the `wkt` and `wkb` keyword args. It's a shame, because I was really looking forward to that convenience. But there are some bad drawbacks.

johnflavin commented:
> I would keep GeoSeries methods and document their usage in cases when your DataFrame comes with wkt and skip GeoDataFrame stuff.

Where would it be best to document this?

martinfleis commented:
> > I would keep GeoSeries methods and document their usage in cases when your DataFrame comes with wkt and skip GeoDataFrame stuff.
> 
> Where would it be best to document this?

We have this example https://geopandas.readthedocs.io/en/latest/gallery/create_geopandas_from_pandas.html which covers WKT. The example itself will soon become Jupyter notebook instead of the current .py file (that is my to-do list for December), so I'd say document it properly in docstrings which go to API now and I'll deal with the example later.

johnflavin commented:
I have removed the `wkb` and `wkt` kwargs from `GeoDataFrame.__init__`.

At the same time, I added the ability to pass a `Series` to `GeoSeries.from_wkb/t`. I thought that added some convenience in that you can do this:
```python
df = pd.DataFrame({'wkt': ['POINT (0 0)'], ...}, index=<could be anything>)
gdf = geopandas.GeoDataFrame(df, geometry=GeoSeries.from_wkt(df['wkt']))
```
To me that makes it simpler to get a `GeoDataFrame` set up from a `DataFrame` with encoded geometries. But that also added a bit of complexity about what to do with the `index` kwarg vs. the index on the `Series`. I opted to raise a `ValueError` if they don't match.

If that complexity isn't worth it, I can revert those changes. But it makes it a little harder to go from wkt df -> gdf; you have to explicitly get the `values` from the wkt `Series`, and if the index is not default you have to set that explicitly too:
```python
df = pd.DataFrame({'wkt': ['POINT (0 0)'], ...}, index=<could be anything>)
gdf = geopandas.GeoDataFrame(df, geometry=GeoSeries.from_wkt(df['wkt'].values, index=df['wkt'].index))
```
At that point you're better off just using `geometry=df['wkt'].apply(shapely.wkt.loads)`. It's less efficient but the syntax is simpler.

martinfleis reviewed (commented):
Thanks a lot for changes! 

We have to support Series, I have left a comment on that in the code.

johnflavin commented:
Thanks for the review!

I've implemented the requested changes. Test failure seems unrelated.

martinfleis commented on geopandas/geodataframe.py:
```diff
@@ -777,6 +789,42 @@ def _to_geo(self, **kwargs):
 
         return geo
 
+    def to_wkb(self):
```
```suggestion
    def to_wkb(self, hex=False):
```

The original array method has `hex` keyword to control. We should expose it here as well. You can copy docstring from pygeos https://pygeos.readthedocs.io/en/latest/io.html#pygeos.io.to_wkb

johnflavin replied:
I had put this in originally, but took it out. I think the reason is that I had also included a `hex` keyword arg in the `from_wkb` methods, but the tests failed because pygeos didn't support loading from wkb hex. Then when I took that out I must have gone a little overboard and removed more than I needed to.

martinfleis commented on geopandas/geodataframe.py:
```diff
@@ -777,6 +789,42 @@ def _to_geo(self, **kwargs):
+
+        # Encode all geometry columns to WKB
+        for col in df.columns[df.dtypes == "geometry"]:
+            df[col] = to_wkb(df[col].values)
```
```suggestion
            df[col] = to_wkb(df[col].values, hex=hex)
```

martinfleis commented on geopandas/geodataframe.py:
```diff
@@ -777,6 +789,42 @@ def _to_geo(self, **kwargs):
+
+        return df
+
+    def to_wkt(self):
```
```suggestion
    def to_wkt(self, **kwargs):
```

Similar as above. The original array method can pass kwargs to pygeos `to_wkt`. We should forward them.

johnflavin replied:
I have done this. I added a note to the docstring that the kwargs are only used if pygeos is installed, stating by implication that they aren't used with shapely. (Since with shapely, `wkt` and `wkb` are properties that can't take any args.) Let me know if that docstring isn't appropriate or consistent with the style or anything like that.

martinfleis replied:
Thanks. I think it is okay. At one point, pygeos will become shapely2.0 so this pygeos/shapely difference is only temporary.

martinfleis commented on geopandas/geodataframe.py:
```diff
@@ -777,6 +789,42 @@ def _to_geo(self, **kwargs):
+
+        # Encode all geometry columns to WKT
+        for col in df.columns[df.dtypes == "geometry"]:
+            df[col] = to_wkt(df[col].values)
```
```suggestion
            df[col] = to_wkt(df[col].values, **kwargs)
```

martinfleis commented on geopandas/geoseries.py:
```diff
@@ -882,6 +979,57 @@ def to_json(self, **kwargs):
+        """
+        return Series(to_wkb(self.array, hex=hex), index=self.index)
+
+    def to_wkt(self):
```
```suggestion
    def to_wkt(self, **kwargs):
```

As above.

martinfleis commented on geopandas/geoseries.py:
```diff
@@ -882,6 +979,57 @@ def to_json(self, **kwargs):
+        --------
+        GeoSeries.to_wkb
+        """
+        return Series(to_wkt(self.array), index=self.index)
```
```suggestion
        return Series(to_wkt(self.array,  **kwargs), index=self.index)
```

martinfleis reviewed (commented):
Thanks a lot for changes. I noticed few more nits.

Test failure is indeed unrelated (something with pandas master we have to look at).

cc @jorisvandenbossche to get a second pair off eyes at this esp. API-wise.

jorisvandenbossche commented on geopandas/geodataframe.py:
```diff
@@ -863,6 +875,57 @@ def _to_geo(self, **kwargs):
+            The default is to return a binary bytes object.
+        kwargs
+            Additional keyword args will be passed to
+            :func:`pygeos.to_wkb` if pygeos is installed.
```
At some point (eg with Shapely 2.0), we should probably just document those keywords

jorisvandenbossche reviewed (approved):
This looks really nice, thanks @johnflavin and @martinfleis

jorisvandenbossche commented:
> We have currently only one function of a similar nature - `geopandas.points_from_xy` which stands a bit out of general API. It feels ideal to have these functions exposed as `GeoSeries.from_wkt()` rather than top-level API as the one for points. 

We should probably think about exposing `points_from_xy` as a `GeoSeries.from_..` constructor as well. One reason that I like it being top-level (and returning an array and not GeoSeries) is that you have less concern about indexes. For example if you currently do:

```
geopandas.GeoDataFrame(
    {'col': [1, 2]}, geometry=geopandas.points_from_xy([1, 2], [3, 4]), index=['a', 'b']
)
```

this currently works. But if `points_from_xy` would return a GeoSeries, the above would no longer work (or at least, you would have to rewrite it to keep it working). 

Now, I think it's certainly still useful to have those `from_..` methods on GeoSeries as well (and the fact that those accept a Series and then preserve the index certainly helps.

jorisvandenbossche commented on geopandas/geoseries.py:
```diff
@@ -360,6 +367,95 @@ def from_file(cls, filename, **kwargs):
+            if index is not None:
+                data = data.reindex(index)
+            data = data.values
+        return cls(from_wkb_or_wkt_function(data, crs=crs), index=index, **kwargs)
```
Just noticed that we should preserve the index of `data` if it's a Series, will push a quick fix.

martinfleis commented:
> Now, I think it's certainly still useful to have those from_.. methods on GeoSeries as well 

Agree. Moved to #1853

martinfleis commented:
Thanks a lot, @johnflavin and @jorisvandenbossche!
